File: javaboxuegu/src/main/java/com/xushengling/javaboxuegu/activity/MainActivity.java
package com.xushengling.javaboxuegu.activity;

//import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
//import androidx.viewpager.widget.PagerAdapter;
//import androidx.viewpager.widget.ViewPager;

import android.content.Intent;
import android.content.SharedPreferences;

import android.os.Bundle;
import android.view.KeyEvent;
//import android.view.LayoutInflater;
import android.view.View;
//import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.xushengling.javaboxuegu.R;
import com.xushengling.javaboxuegu.base.BaseActivity;
import com.xushengling.javaboxuegu.fragment.CourseFragment;
import com.xushengling.javaboxuegu.fragment.ExercisesFragment;
import com.xushengling.javaboxuegu.fragment.MyFragment;

//import java.util.ArrayList;
//import java.util.List;

public class MainActivity extends BaseActivity implements View.OnClickListener {
//    View view1, view2, view3;
//    ViewPager viewPager;
//    List<View> viewList;
    FrameLayout mBodyLayout;
    /**
     * 底部按钮栏
     */
    private LinearLayout mButtonLayout;

    /**
     * 底部按钮
     */
    private View mCuresBtn;
    private View mExercisesBtn;
    private View mMyInfoBtn;
    private TextView tv_course;
    private TextView tv_exercise;
    private TextView tv_myInfo;
    private ImageView iv_course;
    private ImageView iv_exercise;
    private ImageView iv_myInfo;
    private TextView tv_main_titer;
    private ConstraintLayout rl_title_bar;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initBottomBtn();
        setListener();
        setInitStarts();
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_main;
    }

    @Override
    public void initView() {
//        initViewPager();
        TextView tv_back = $(R.id.backTv);
        tv_main_titer = $(R.id.titleTv);
        tv_main_titer.setText(getText(R.string.main_activity));
        rl_title_bar = $(R.id.mainTitleBar);
        rl_title_bar.setBackgroundColor(this.getColor(R.color.colorBar));
        tv_back.setVisibility(View.GONE);
        initBodyLayout();
    }

//    private void initViewPager() {
//        viewPager=findViewById(R.id.viewpager);
//        LayoutInflater inflater=getLayoutInflater();
//        view1=inflater.inflate(R.layout.course_fragment,null);
//        view2=inflater.inflate(R.layout.exercises_fragment,null);
//        view3=inflater.inflate(R.layout.my_fragment,null);
//
//        viewList=new ArrayList<>();
//        viewList.add(view1);
//        viewList.add(view2);
//        viewList.add(view3);
//
//        PagerAdapter pagerAdapter=new PagerAdapter() {
//            @Override
//            public int getCount() {
//                return viewList.size();
//            }
//
//            @Override
//            public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
//                return view==object;
//            }
//
//            @NonNull
//            @Override
//            public Object instantiateItem(@NonNull ViewGroup container, int position) {
//                container.addView(viewList.get(position));
//                return viewList.get(position);
//            }
//
//            @Override
//            public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
//                container.removeView(viewList.get(position));
//            }
//        };
//        viewPager.setAdapter(pagerAdapter);
//    }

    private void initBottomBtn() {
        mButtonLayout=$(R.id.main_btn_bar);
        mCuresBtn=$(R.id.main_btn_bar_course_ll);
        mExercisesBtn=$(R.id.main_btn_bar_exercises_ll);
        mMyInfoBtn=$(R.id.main_btn_bar_my_ll);
        tv_course=$(R.id.main_btn_bar_course_tv);
        tv_exercise=$(R.id.main_btn_bar_exercises_tv);
        tv_myInfo=$(R.id.main_btn_bar_my_tv);
        iv_course=$(R.id.main_btn_bar_course_btn);
        iv_exercise=$(R.id.main_btn_bar_exercises_btn);
        iv_myInfo=$(R.id.main_btn_bar_my_btn);
    }
    private void initBodyLayout(){
        mBodyLayout=$(R.id.main_body);
        replaceFragment(new CourseFragment());
    }

    @Override
    public void onClick(View view) {
    switch (view.getId()){
        case R.id.main_btn_bar_course_ll:
            clearBottomImageState();
            selectDisplayView(0);
            break;
        case R.id.main_btn_bar_exercises_ll:
            clearBottomImageState();
            selectDisplayView(1);
            break;
        case R.id.main_btn_bar_my_ll:
            clearBottomImageState();
            selectDisplayView(2);
            break;
    }
    }
    /**
     * 设置底部三个按钮的监听事件
     */
    private void setListener(){
        for(int i = 0 ; i <mButtonLayout.getChildCount(); i++){
            mButtonLayout.getChildAt(i).setOnClickListener(this);
        }
    }

    /**
     * 清楚底部选中的按钮
     */
    private void clearBottomImageState(){
        tv_course.setTextColor(this.getColor(R.color.colorText));
        tv_exercise.setTextColor(this.getColor(R.color.colorText));
        tv_myInfo.setTextColor(this.getColor(R.color.colorText));
        iv_course.setImageResource(R.drawable.main_course_icon);
        iv_exercise.setImageResource(R.drawable.main_exercises_icon);
        iv_myInfo.setImageResource(R.drawable.main_my_icon);

    }

    /**
     * 设置底部按钮的选中状态
     */
    private void setSelectedStatus(int index){
        switch (index){
            case 0:
                mCuresBtn.setSelected(true);
                iv_course.setImageResource(R.drawable.main_course_icon_selected);
                tv_course.setTextColor(this.getColor(R.color.colorTextSelected));
                rl_title_bar.setVisibility(View.VISIBLE);
                tv_main_titer.setText(R.string.main_activity);
                break;
            case 1:
                mExercisesBtn.setSelected(true);
                iv_exercise.setImageResource(R.drawable.main_exercises_icon_selected);
                tv_exercise.setTextColor(this.getColor(R.color.colorTextSelected));
                rl_title_bar.setVisibility(View.VISIBLE);
                tv_main_titer.setText(R.string.main_exercises_tv);
                break;
            case 2:
                mMyInfoBtn.setSelected(true);
                iv_myInfo.setImageResource(R.drawable.main_my_icon_selected);
                tv_myInfo.setTextColor(this.getColor(R.color.colorTextSelected));
                rl_title_bar.setVisibility(View.GONE);
                break;
        }
    }

    /**
     * 设置界面view的初始化状态
     */
    private void setInitStarts(){
        clearBottomImageState();
        setSelectedStatus(0);
        createView(0);
    }
    /**
     * 显示对应的界面
     */
    private void selectDisplayView(int index){

        createView(index);
        setSelectedStatus(index);
    }
    /**
     * 选择视图
     */
    private void createView(int viewIndex){
        switch (viewIndex){
            case 0:
                //课程界面
                replaceFragment(new CourseFragment());
                break;
            case 1:
                //习题界面
                replaceFragment(new ExercisesFragment());
                break;
            case 2:
                //我的界面
                replaceFragment(new MyFragment());
                break;
        }
    }
    protected long exitTime;//记录第一次点击时的时间

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_DOWN){
            if ((System.currentTimeMillis() - exitTime) > 2000){
               Toast(getString(R.string.keyDown));
                exitTime = System.currentTimeMillis();
            }else { this.finish();
            if (readLoginStatus()){
                clearLoginStatus();
            }
            System.exit(0);
            }
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }
    /**
     * 获取 SharedPreferences 中的登录信息
     */
    private boolean readLoginStatus(){
        SharedPreferences sp=getSharedPreferences("LoginInfo",MODE_PRIVATE);
        return sp.getBoolean("isLogin",false);
    }
    /**
     * 清除 SharedPreferences 中的登录状态
     */

    private void clearLoginStatus(){
        SharedPreferences sp=getSharedPreferences("LoginInfo",MODE_PRIVATE);
        SharedPreferences.Editor editor=sp.edit();
        editor.putBoolean("isLogin",false);
        editor.putString("loginUserName","");
        editor.apply();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (data != null){
            //从设置界面或登录界面传递过来的登陆状态
            boolean isLogin=data.getBooleanExtra("isLogin",false);
            if (isLogin){
                //登陆成功界面
                clearBottomImageState();
                selectDisplayView(0);
            }
        }
    }
    /**
     * 修改 Fragment界面
     */
    private void replaceFragment(Fragment fragment){
        FragmentManager fragmentManager=getSupportFragmentManager();
        FragmentTransaction transaction=fragmentManager.beginTransaction();
        transaction.replace(R.id.main_body,fragment);
        transaction.commit();
    }
}

File: app/src/main/java/com/xushengling/myapplication/activity/MainActivity.kt
package com.xushengling.myapplication.activity

import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import android.widget.LinearLayout
import android.widget.Toast
import android.widget.Toast.LENGTH_SHORT
import androidx.appcompat.app.AppCompatActivity
import com.xushengling.myapplication.R
import com.xushengling.myapplication.view.MyInfoView
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.main_title_bar.*
import kotlin.system.exitProcess

class MainActivity : AppCompatActivity(),View.OnClickListener{
    /**
     * 底部按钮栏
     */
    private lateinit var mButtonLayout: LinearLayout
    private var mMyInfoView:MyInfoView? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initView()
        initBottomBtn()
        setListener()
        setInitStarts()
    }

    private fun initBottomBtn() {
        mButtonLayout = findViewById(R.id.main_btn_bar)

    }

    private fun setInitStarts() {
        clearButtonImageState()
        setSelectedStatus(0)
        createView(0)
    }
    /**
     * 设置底部三个按钮的监听事件
     */
    private fun setListener() {
        for (i in 0 until mButtonLayout.childCount) {
            mButtonLayout.getChildAt(i).setOnClickListener(this)
        }
    }

    /**
     * 移除不需要视图
     */
    private fun removeAllView() {
        for (i in 0 until main_body.childCount) {
            main_body.getChildAt(i).visibility = View.GONE
        }
    }

    private fun initView() {
        mainTitleBar.setBackgroundColor(Color.parseColor("#30B4FF"))
        backTv.visibility = View.GONE
    }

    override fun onClick(v: View?) {
        when(v!!.id){
            R.id.main_btn_bar_course_ll ->{
                clearButtonImageState()
                selectDisplayView(0)
            }
            R.id.main_btn_bar_exercises_ll ->{
                clearButtonImageState()
                selectDisplayView(1)
            }
            R.id.main_btn_bar_my_ll ->{
                clearButtonImageState()
                selectDisplayView(2)
            }
        }
    }

    /**
     * 显示对应的界面
     */
    private fun selectDisplayView(index: Int) {
        removeAllView()
        createView(index)
        setSelectedStatus(index)
    }

    /**
     * 清楚底部选中的按钮
     */
    private fun clearButtonImageState() {
        val color=getColor(R.color.colorText)
        main_btn_bar_course_tv.setTextColor(color)
        main_btn_bar_exercises_tv.setTextColor(color)
        main_btn_bar_my_tv.setTextColor(color)
        main_btn_bar_course_btn.setImageResource(R.drawable.main_course_icon)
        main_btn_bar_exercises_btn.setImageResource(R.drawable.main_exercises_icon)
        main_btn_bar_my_btn.setImageResource(R.drawable.main_my_icon)
        for (i in 0 until mButtonLayout.childCount) {
            mButtonLayout.getChildAt(i).isSelected = false
        }
    }


    private fun setSelectedStatus(index: Int) {
        when(index){
            0 ->{
                main_btn_bar_course_ll.isSelected = true
                main_btn_bar_course_btn.setImageResource(R.drawable.main_course_icon_selected)
                main_btn_bar_course_tv.setTextColor(Color.parseColor("#0097F7"))
                mainTitleBar.visibility=View.VISIBLE
                titleTv.text = "博学谷课程"
            }
            1 ->{
                main_btn_bar_exercises_ll.isSelected = true
                main_btn_bar_exercises_btn.setImageResource(R.drawable.main_exercises_icon_selected)
                main_btn_bar_exercises_tv.setTextColor(Color.parseColor("#0097F7"))
                mainTitleBar.visibility=View.VISIBLE
                titleTv.text = "博学谷习题"
            }
            2 ->{
                main_btn_bar_my_ll.isSelected = true
                main_btn_bar_my_btn.setImageResource(R.drawable.main_my_icon_selected)
                main_btn_bar_my_tv.setTextColor(Color.parseColor("#0097F7"))
                mainTitleBar.visibility=View.GONE
            }
        }
    }

    private fun createView(viewIndex: Int) {
        when(viewIndex){
            0 ->{
                //课程界面
            }
            1 ->{
                //习题界面
            }
            2 ->{
                //我的界面
                if (mMyInfoView == null) main_body.addView(MyInfoView(this).view)
                else mMyInfoView!!.view

            }
        }
    }

    private var exitTime:Long = 0
    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK && event.action == KeyEvent.ACTION_DOWN){
            if ((System.currentTimeMillis() - exitTime) > 2000){
                toast("再次一次退出博学谷")
                exitTime=System.currentTimeMillis()
            }else{
                this.finish()
                if (readLoginStatus()){
                    clearLoginStatus()
                }
                exitProcess(0)
            }
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    /**
     * 获取 SharedPreference 中的登录情况
     */
    private fun readLoginStatus():Boolean {
        return  getSharedPreferences("loginInfo", Context.MODE_PRIVATE).getBoolean("isLogin",false)
    }

    /**
     * 清除 SharedPreference 中的登陆状态
     */
    private fun clearLoginStatus(){
        val sp=getSharedPreferences("loginInfo", Context.MODE_PRIVATE)
        val editor=sp.edit()
        editor.putBoolean("isLogin",false)
        editor.putString("loginUserName","")
        editor.apply()
        editor.commit()
    }

    private fun toast(value:String){
        Toast.makeText(this,value,LENGTH_SHORT).show()
    }

}

